flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderRetrievalServiceTest.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.highavailability;
+
+import org.apache.flink.kubernetes.kubeclient.FlinkKubeClient;
+import org.apache.flink.kubernetes.kubeclient.resources.KubernetesConfigMap;
+import org.apache.flink.kubernetes.utils.Constants;
+
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for the {@link KubernetesLeaderRetrievalService}.
+ */
+public class KubernetesLeaderRetrievalServiceTest extends KubernetesHighAvailabilityTestBase {
```

Comment:
What about integration tests for the `KubernetesLeaderElectionService` and the `KubernetesLeaderRetrievalService` with some K8s server?

Reply:
Just like Yarn MiniCluster, we are not very simple to start a minikube. So I prefer to add the unit tests for the contract testing first and leave the integration tests with E2E tests together. WDYT?

Reply:
Hmm, I think it would be good to actually test the ha services against some form of K8s testing server/cluster. I mean they can run as part of the E2E tests but we should not only test a full Flink cluster with K8s ha enabled.

Reply:
Yes. You are right. What I mean is that we could also add some individual tests for HA service, not just a full Flink cluster with K8s ha enabled. They will be implemented in the E2E module. I have left a comment in FLINK-19545.

Reply:
Shouldn't these tests be part of the commits where we add the K8s services? That way it will be a bit easier to review them for completeness.